File: leetcode/07-Generate-Parenthness/python/main.py
class Solution:
    def generateParenthesis(self, n: int):
        result = []
        # self.tmp(n, "", 0, 0, result)
        self.another(n, 0, 0, "", result)
        return result

    def tmp(self, n, s, counter, total, result):
        if len(s) == 2*n:
            result[s] = ""
            return

        if counter == 0:
            self.tmp(n, s + "(", counter + 1, total + 1, result)
        else:
            self.tmp(n, s + ")", counter - 1, total, result)

        if total < n:
            self.tmp(n, s + "(", counter + 1, total + 1, result)

    def another(self, n, left, right, s, result):
        if len(s) == 2*n:
            result.append(s)
            return

        if left < n:
            self.another(n, left+1, right, s + "(", result)
        if right < left:
            self.another(n, left, right+1, s + ")", result)

File: leetcode/07-Generate-Parenthness/go/main.go
package main

import "slices"

func generateParenthesis(n int) []string {
	result := make([]string, 0)

	genetate(n, 0, 0, "", &result)

	return result
}

func genetate(n int, left int, right int, s string, result *[]string) {
	if len(s) == 2*n && !slices.Contains(*result, s) {
		*result = append(*result, s)
		return
	}

	if left > right {
		genetate(n, left, right+1, s+")", result)
	}
	if left < n {
		genetate(n, left+1, right, s+"(", result)
	}
}
